### PR TITLE
storage: fix send/recv unprotected when 2 at once

### PR DESCRIPTION
**Pull request by Gerold103**
The PR is for #434, and the first commit is cherry-picked from #439 to speed up the tests.

### CODE REVIEW

File: test/storage-luatest/storage_1_1_test.lua [hunk 1]
```diff
@@ -261,3 +261,87 @@ test_group.test_on_bucket_event = function(g)
...
+--
+-- gh-434: bucket_recv() shouldn't change the transfer flags if a transfer for
+-- the same bucket is already in progress.
+--
+test_group.test_bucket_double_recv = function(g)
+    g.replica_2_a:exec(function(bid, uuid)
+        local transfer_flags =
+            ivshard.storage.internal.rebalancer_transfering_buckets
+        local f = ifiber.create(ivshard.storage.bucket_recv, bid, uuid, {})
+        f:set_joinable(true)
```
Comment (Serpentian) [inline test/storage-luatest/storage_1_1_test.lua:326]: Happens rarely:
```
[002] storage-luatest/storage_1_1_test.lua
[002] TAP version 13
[002] 1..2
[002] # Started on Tue Oct 31 21:11:03 2023
[002] # Starting group: storage
[002] ok     1	storage.test_bucket_double_send
[002] not ok 2	storage.test_bucket_double_recv
[002] #   ...ing/tnt/vshard/test/storage-luatest/storage_1_1_test.lua:326: the fiber is dead
[002] #   stack traceback:
[002] #   	...ing/tnt/vshard/test/storage-luatest/storage_1_1_test.lua:322: in function 'storage.test_bucket_double_recv'
[002] #   	...
[002] #   	[C]: in function 'xpcall'
[002] # Ran 2 tests in 0.327 seconds, 1 success, 1 error
```

File: vshard/storage/init.lua [hunk 2]
```diff
@@ -2155,14 +2173,18 @@ local function bucket_send(bucket_id, destination, opts)
     if type(bucket_id) ~= 'number' or type(destination) ~= 'string' then
         error('Usage: bucket_send(bucket_id, destination)')
     end
-    M.rebalancer_transfering_buckets[bucket_id] = true
+    local ret
```
Comment (Serpentian) [inline vshard/storage/init.lua:2176]: Nit: let's make the code consistent with `bucket_recv`:

```diff
diff --git a/vshard/storage/init.lua b/vshard/storage/init.lua
index fffe6a1..899f6f0 100644
--- a/vshard/storage/init.lua
+++ b/vshard/storage/init.lua
@@ -2173,8 +2173,8 @@ local function bucket_send(bucket_id, destination, opts)
     if type(bucket_id) ~= 'number' or type(destination) ~= 'string' then
         error('Usage: bucket_send(bucket_id, destination)')
     end
-    local ret
-    local status, err = bucket_transfer_start(bucket_id)
+    local status, ret, err
+    status, err = bucket_transfer_start(bucket_id)
     if not status then
         return nil, err
     end

```

File: test/storage-luatest/storage_1_1_test.lua [hunk 3]
```diff
@@ -261,3 +261,87 @@ test_group.test_on_bucket_event = function(g)
         box.space.data2:drop()
     end)
 end
+
```
Comment (Serpentian) [inline test/storage-luatest/storage_1_1_test.lua:264]: Please, fix checkpatch warning in commit msg

File: test/storage-luatest/storage_1_1_test.lua [hunk 4]
```diff
@@ -261,3 +261,87 @@ test_group.test_on_bucket_event = function(g)
...
+        ilt.assert(f_ok)
+        ilt.assert(ok)
+        ilt.assert(not transfer_flags[bid])
+        _G.bucket_gc_wait()
+    end, {bid})
+    --
+    -- Cleanup.
+    --
+    g.replica_2_a:exec(function(bid, uuid)
+        local ok, err = ivshard.storage.bucket_send(bid, uuid,
```
[the remark this answers is not in the document]
Comment (Gerold103) [reply test/storage-luatest/storage_1_1_test.lua:309]: The failures you mention in all the comments are not related to the patch. They are from separate bugs which I've fixed in another PR (which awaits your review already): https://github.com/tarantool/vshard/pull/439/commits/95971572d6b1cae829d15aa558c7c18d5f27f3b9 and https://github.com/tarantool/vshard/pull/439/commits/708070d480ae08e5d668e6ace3112518ffc079f7.

File: vshard/storage/init.lua (hunk 2 above)
Comment (Gerold103) [reply vshard/storage/init.lua]: Applied.

File: test/storage-luatest/storage_1_1_test.lua (hunk 3 above)
Comment (Gerold103) [reply test/storage-luatest/storage_1_1_test.lua:264]: Done.

File: test/storage-luatest/storage_1_1_test.lua (hunk 1 above)
Comment (Gerold103) [reply test/storage-luatest/storage_1_1_test.lua:326]: I couldn't reproduce it. It either passes or fails earlier on one of the bugs which I've fixed in another PR. But I did apply this:
```Diff
+++ b/test/storage-luatest/storage_1_1_test.lua
@@ -341,6 +341,7 @@ test_group.test_bucket_double_recv = function(g)
         -- Cleanup.
         --
         _G.bucket_recovery_wait()
+        _G.bucket_gc_wait()
         ilt.assert_equals(box.space._bucket:get{bid}, nil)
     end, {vtest.storage_first_bucket(g.replica_1_a),
           g.replica_1_a:replicaset_uuid()})
```
The recovery turns the bucket into GARBAGE, and it sometimes affects runs of newer groups with other configs.

**Review by Serpentian — APPROVED** [review]
Thank you for the patch! Applied it on top of auto-master patchset, tests, added in this PR, are 100% stable now.

P.S. I'll add review to the auto-master by the end of a week

**Pull request merged**